### DISCUSSION TITLE
yaml titles vs first-level headers in quarto pdf books

### DISCUSSION THREAD

**Discussion by d-morrison** [quarto-dev/quarto-cli#6619] [Q&A] [answered]
### Description

I'm building a quarto book (https://github.com/d-morrison/rme, https://d-morrison.github.io/rme/), and I'm having trouble with the header numbering in the pdf version (https://d-morrison.github.io/rme/Regression-Models-for-Epidemiology.pdf).

Currently, I've defined each chapter's title in the yaml header for the corresponding .qmd file, and I've used level-one headers (`#`) to define sections within the chapter. This works fine for the html version of the book (https://d-morrison.github.io/rme/) as long as I set `shift-heading-level-by: 1` in the project-level `.yml` file (https://github.com/d-morrison/rme/blob/main/_quarto-book.yml).

However, the pdf version is not working correctly: it treats the chapter yaml title as a level-one header, and whether I set `shift-heading-level-by: 1` or `shift-heading-level-by: 0`, I still get the chapter title and the level-one-headers treated as equivalent.

Am I configuring the pdf settings incorrectly, or is this a bug?

Thanks,
Ezra

**Comment by mcanouil** [comment]
FYI, I am going to look into this later this week (if someone else does not in the meantime).

**Comment by cderv** [comment]
> Currently, I've defined each chapter's title in the yaml header for the corresponding .qmd file, and I've used level-one headers (#) to define sections within the chapter.

In our doc, we explain that title for chapters can be either level 1 heading using `#` or defined using `title:` YAML field. So it is highly possible that we treat both the same way, and that `shift-heading-level-by` applies on it. Though I understand that it works somehow for HTML and not for PDF. 

So I am curious of why doing that, before we can see how `shift-heading-level-by` play with books. Why not use `##` and `###` and keep `#` for title ?

**Reply by d-morrison** [reply]
Thanks for looking into it! I think I could use ## and ### as you suggest instead of `shift-heading-level-by`;  if I were starting from scratch I would probably just do that, but given how much I've already written, it's a bit tedious to demote all the existing sections by hand. I guess I could use a regex find-replace all, but just a standard find-replace of # with ## would turn ## into #### instead of ###. Using # for chapter titles also means typing an extra character for each heading going forward, which of course is no big deal, just slightly annoying and less minimalist.

**Reply by cscheid** [reply]
> but just a standard find-replace of # with ## 

If your text editor supports searching by regexp, `^#` will only catch `#` starting at the beginning of the line. I've found that to be a useful trick in the past.

**Comment by mcanouil** [comment]
@d-morrison Did you manage to get the desired results?

**Reply by d-morrison** [reply]
Apologies for the very delayed reply! Switching chapter titles to first-level headers worked for the pdf book. However, I'm also using each chapter as a revealjs slide deck (using quarto profiles), and that doesn't work well with the chapter/deck title being a first-level header; even with `slide-level: 3` in the yaml header, `##` aren't being treated as sections. I built a reprex here: https://github.com/d-morrison/quarto.test.book

# with chapter titles in the yaml header (branch [`intro-slides-work`](https://github.com/d-morrison/quarto.test.book/tree/intro-slides-work))

- Here's the outline view of `intro.html` with chapter titles in the yaml header, compiled with `quarto render --profile=revealjs` with `slide-level: 2` in `_quarto-book.yml` (this is as expected):

![image](https://github.com/quarto-dev/quarto-cli/assets/2474437/6e2e7669-2b18-47df-b8bc-8cb0ec604276)

- But here's the TOC for the pdf book, with chapter titles in the yaml header, compiled with `quarto render --profile=book` and `shift-heading-level-by: 0` (this is not as expected: sections are being promoted to chapters):

![image](https://github.com/quarto-dev/quarto-cli/assets/2474437/69323375-fbc3-412e-8900-8a8703655a44)

- if we set `shift-heading-level-by: 1` (branch [`intro-slides-work-pdf-shift-heading-1`](https://github.com/d-morrison/quarto.test.book/tree/intro-slides-work-shift-heading-1)), it's still broken (this is not as expected: chapter titles are being demoted to sections):

![image](https://github.com/quarto-dev/quarto-cli/assets/2474437/372680e5-4a1e-4ae9-8afd-bbbe61aec559)

# with chapter titles as first-level headers  (branch [`intro-slides-dont-work`](https://github.com/d-morrison/quarto.test.book/tree/intro-slides-dont-work)):

- Here's the outline view of `intro.html` with chapter titles as first-level headers, compiled with `quarto render --profile=revealjs` with `slide-level: 3` in `_quarto-book.yml` (not as expected; compare with version above):

![image](https://github.com/quarto-dev/quarto-cli/assets/2474437/4d9f5b68-55d4-41ee-8056-54f8afd74c4f)

- here's the TOC for the pdf book, with chapter titles as first-level headers, compiled with `quarto render --profile=book` (this is as expected):

![image](https://github.com/quarto-dev/quarto-cli/assets/2474437/bc75b93c-8d1c-4550-bc96-d1c7fd03d714)

**Reply by mcanouil** [reply, marked as answer]
Level 1 header is equivalent to `title` so it's exactly as expected.

Starts your section from a level 2, and use `shift-heading-level-by: -1` for `revealjs`.

PS: when sharing an example repository, please keep it simple: no need for code cells or any computations and no need for 10+ unrelated options, which only makes it harder to focus on what matters.

**Reply by d-morrison** [reply]
Thanks! That worked. Apologies for the messy reprex!

**Reply by d-morrison** [reply]
One more minor issue/question; some of my book chapters have sub-titles defined in their yaml headers; for example:

````
---
title: "Models for Count Outcomes"
subtitle: "Poisson regression and variations"
---
````

<img width="525" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/2474437/abcd5bf2-f88c-4d8c-8d18-5c25a3fbb4eb">

Is there any way to keep those sub-titles  using the title = # format?

**Reply by mcanouil** [reply]
I don't follow, keep where?

The "title = first heading" thing is only in this context as a document needs to have a title.
There is no alternative to `subtitle`. only first heading is promoted to title when there is none. There are no other promotion for headings.

<details><summary>A not working workaround since metadata are gathered too early.</summary>

Note that you can use conditional contents to make particular conditioning: https://quarto.org/docs/authoring/conditional.html

````qmd
::: {.content-visible when-format="html" when-format="pdf" unless-format="revealjs"}

---
title: "Models for Count Outcomes"
subtitle: "Poisson regression and variations"
---

:::

::: {.content-visible when-format="revealjs"}

# Models for Count Outcomes

:::

````

</details>

**Reply by cscheid** [reply]
@mcanouil No, that doesn't work; metadata is lifted out of content before conditional content is evaluated. This is a long-standing bug in Quarto+Pandoc that we can't easily fix.

**Reply by mcanouil** [reply]
Indeed, not all metadata can work like that.
I update my reply, as only the workaround does not work.
